File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n);
        vector<int> pre(n);
        vector<int> suf(n);
        int pro = 1;
        pre[0] = 1;
        suf[n-1] = 1;
        pro = nums[0];
        for(int i=1; i<n; i++){
            pre[i] = pro;
            pro *= nums[i];
        }
        pro = nums[n-1];
        for(int i=n-2; i>=0; i--){
            suf[i] = pro;
            pro *= nums[i];
        }

        for(int i=0; i<n; i++){
            ans[i] = pre[i] * suf[i];
        }
        return ans;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        if len(nums) == 0:
            return []

        pre = [1] * len(nums)
        suf = [1] * len(nums)

        prod = nums[0]
        for idx in range(1, len(nums)):
            pre[idx] = prod
            prod *= nums[idx]

        prod = nums[-1]
        for idx in range(len(nums)-2, -1, -1):
            suf[idx] = prod
            prod *= nums[idx]

        ans = []
        for idx in range(len(nums)):
            ans.append(pre[idx] * suf[idx])
        
        return ans
        